sdk/dotnet/Pulumi.Tests/Deployment/DeploymentRunnerTests.cs

```diff
@@ -0,0 +1,174 @@
+// Copyright 2016-2021, Pulumi Corporation
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+using Microsoft.Extensions.Logging;
+using Xunit;
+
+using Pulumi.Testing;
+using Pulumi.Tests.Mocks;
+
+namespace Pulumi.Tests
+{
+    public class DeploymentRunnerTests
+    {
+        // Regression test. This used to fail with
+        // "System.Collections.Generic.KeyNotFoundException: The given
+        // key 'System.Threading.Tasks.Task+DelayPromise' was not
+        // present in the dictionary."
+        [Fact]
+        public async Task WorksUnderStress()
+        {
+            await Deployment.TestAsync<StressRunnerStack>(new EmptyMocks());
+        }
+
+        class StressRunnerStack : Stack
+        {
+            public StressRunnerStack()
+            {
+                var runner = Pulumi.Deployment.Instance.Internal.Runner;
+
+                for (var i = 0; i < 100; i++)
+                {
+                    runner.RegisterTask($"task{i}", Task.Delay(100 + i));
+                }
+
+                // This introduces a race between two simultaneous
+                // `WhileRunningAsync`, the second one being implicit
+                // in runing this stack, and is necessary for the
+                // repro.
+                ((Deployment.Runner)runner).WhileRunningAsync();
```

Comment:
Is that even legal to do so? Should we prohibit this? Is there no way to repro in terms of user API?

(Sorry if I missed your earlier comment about that)

Reply:
Yes it's illegal. This can't happen. I could not repro this using legal code unfortunately. Let me just drop this test and check the changes in with no repro of the original problem. I have some indirect indication that the changes are helping (`dotnet test` becomes stable on my machine).